### PR TITLE
Fix issue #557 (libdeepspeech performance)

### PR DESCRIPTION
**Pull request by Cwiiis**

### CODE REVIEW

File: native_client/deepspeech.h [hunk 1]
```diff
@@ -4,83 +4,119 @@
...
-                       int aSampleRate,
-                       float** aMfcc,
-                       int* aNFrames = NULL,
-                       int* aFrameLen = NULL);
+    public:
+      /**
+       * @brief Initialise a DeepSpeech context.
+       *
+       * @param aModelPath The path to the frozen model graph.
+       * @param aNCep The number of cepstrum the model was trained with.
```
Comment (lissyx) [inline native_client/deepspeech.h]: At some point, I was thinking: what about storing that and ncontext within a tensor in the model? Like we do for `input_lenghts` with `get_tensor_by_name()` ?

File: native_client/deepspeech.h [hunk 2]
```diff
@@ -4,83 +4,119 @@
...
-                int aNFrames,
-                int aFrameLen = 0);
+      /**
+       * @brief Given audio, return a vector suitable for input to the
+       *        DeepSpeech model.
+       *
+       * Extracts MFCC features from a given audio signal and adds the
+       * appropriate amount of context to run inference on the DeepSpeech model.
+       *
+       * @param aBuffer A 16-bit, mono raw audio signal at the appropriate
```
Comment (lissyx) [inline native_client/deepspeech.h]: Would it make sense we state that "appropriate" is what the model has been trained against ?

File: native_client/deepspeech.h [hunk 3]
```diff
@@ -4,83 +4,119 @@
...
+       *         responsible for freeing this string.
+       */
+      char* infer(float* aMfcc,
+                  int aNFrames,
+                  int aFrameLen = 0);
+
+      /**
+       * @brief Use the DeepSpeech model to perform Speech-To-Text.
+       *
+       * @param aMfcc An MFCC features array.
```
Comment (lissyx) [inline native_client/deepspeech.h]: There is no such `aMfcc` parameter there :)

File: native_client/deepspeech.h (hunk 1 above)
Comment (Cwiiis) [inline native_client/deepspeech.h]: If we stored that in a tensor, I guess we'd also have to use Tensorflow to process the audio data, which probably wouldn't be great for parallelism(?) (or maybe it'd be better that way?) - I think this is outside of the scope of this patch though.

File: native_client/deepspeech.h (hunk 2 above)
Comment (Cwiiis) [inline native_client/deepspeech.h]: I think this is explicit enough :)

File: native_client/deepspeech.h (hunk 3 above)
Comment (Cwiiis) [inline native_client/deepspeech.h]: Good catch :)

File: native_client/BUILD [hunk 4]
```diff
@@ -1,30 +1,62 @@
...
-    srcs = ["deepspeech.cc"],
-    hdrs = ["deepspeech.h"],
-    deps = ["//tensorflow/core:core",
-            ":c_speech_features"]
-)
+cc_binary(
+    name = "libdeepspeech.so",
+    srcs = ["deepspeech.cc", "deepspeech.h"],
+    deps = [":c_speech_features",
+            "//tensorflow/core:core_cpu",
```
Comment (lissyx) [inline native_client/BUILD]: nit: might be inconsistent wrt space/tabs

File: native_client/BUILD [hunk 5]
```diff
@@ -1,30 +1,62 @@
...
+            "//tensorflow/core:direct_session",
+            "//tensorflow/core/kernels:logging_ops",
```
Comment (lissyx) [inline native_client/BUILD]: nit: same

File: native_client/BUILD (hunk 4 above)
Comment (Cwiiis) [inline native_client/BUILD]: Double-checking here, there are only spaces and it lines up correctly.

Comment (lissyx) [inline native_client/BUILD]: Plus, "core_cpu" limits ourselves to no GPU support. Is that what we really want?

File: native_client/BUILD (hunk 5 above)
Comment (Cwiiis) [inline native_client/BUILD]: I think this is Github being weird, it all lines up correctly.

File: tc-package.sh [hunk 6]
```diff
@@ -4,13 +4,9 @@ set -xe
 
 mkdir -p /tmp/artifacts
 
-tar -C ${HOME}/DeepSpeech/tf/bazel-bin/tensorflow/ \
```
Comment (lissyx) [inline tc-package.sh]: Does it means libtensorflow.so content gets into libdeepspeech.so ?

Comment (Cwiiis) [inline tc-package.sh]: Right, libdeepspeech.so is self-contained after this patch.

File: native_client/BUILD [hunk 7]
```diff
@@ -1,30 +1,62 @@
...
+#            "//tensorflow/core/kernels:resource_variable_ops",
+#            "//tensorflow/core/kernels:sdca_ops",
+#            "//tensorflow/core/kernels:set_kernels",
+            "//tensorflow/core/kernels:sparse",
+#            "//tensorflow/core/kernels:state",
+#            "//tensorflow/core/kernels:string",
+#            "//tensorflow/core/kernels:training_ops",
+#            "//tensorflow/core/kernels:word2vec_kernels",
+    ],
+    linkshared = 1,
```
Comment (lissyx) [inline native_client/BUILD]: That makes a huge set of stuff to maintain.

Comment (Cwiiis) [inline native_client/BUILD]: True, I'm not sure I really like this part of the patch - I'm willing to just go with all_kernels, but the space saving is quite nice.

File: native_client/BUILD (hunk 4 above)
Comment (Cwiiis) [inline native_client/BUILD]: I think for running inference, yes, pretty much - Or at least, we should special case things and perhaps have a separate target for with GPU support if we really want it. Given the limited set of GPUs supported and the intended use-case for the library, I'd question how useful the GPU support is.

File: tc-package.sh (hunk 6 above)
Comment (lissyx) [inline tc-package.sh]: I have to admit I liked it a bit better with just libtensorflow.so + libdeepspeech.so: one would be able to replace one piece without having to rebuild all.

File: native_client/c_speech_features/c_speech_features.c [hunk 8]
```diff
@@ -13,30 +13,39 @@ csf_mfcc(const short* aSignal, unsigned int aSignalLen, int aSampleRate,
...
   csf_float* feat;
   csf_float* energy;
 
   int n_frames = csf_logfbank(aSignal, aSignalLen, aSampleRate, aWinLen, aWinStep,
                               aNFilters, aNFFT, aLowFreq, aHighFreq, aPreemph,
                               aWinFunc, &feat, aAppendEnergy ? &energy : NULL);
 
+  // Allocate an array so we can calculate the inner loop multipliers
+  // in the DCT-II just one time.
+  double* dct2f = (double*)malloc(sizeof(double) * aNFilters * aNCep);
```
Comment (lissyx) [inline native_client/c_speech_features/c_speech_features.c]: Is that the trick that prevented proper optimization ?

File: tc-package.sh (hunk 6 above)
Comment (Cwiiis) [inline tc-package.sh]: You wouldn't need to rebuild it all for code changes, just the files that change (and any dependents) - In terms of distribution, it would allow us to distribute libdeepspeech.so changes outside of libtensorflow.so, which would be nice. Static linking in a Bazel library build isn't supported though, so we'd either have to live with the gnuc99 warning or distribute libdeepspeech and libc_speech_features separately.

On the other hand, I do like that this is just one file to worry about installing and distributing, and I would suppose there's the possibility that there's a system libtensorflow.so that we might conflict with. Pros and cons :)

File: native_client/c_speech_features/c_speech_features.c (hunk 8 above)
Comment (Cwiiis) [inline native_client/c_speech_features/c_speech_features.c]: This is the part that meant that mfcc wasn't performing as well as it should. This didn't affect the optimisation (that was the disabled flags in the BUILD file).

File: native_client/README.md [hunk 9]
```diff
@@ -18,18 +18,12 @@ ln -s ../DeepSpeech/native_client ./
...
 
 ```
-bazel build -c opt //tensorflow:libtensorflow.so
-```
-
-Then you can build the DeepSpeech native library.
-
-```
-bazel build -c opt //native_client:*
+bazel build -c opt --copt=march=native --copt=mtune=native --copt=-O3 //tensorflow:libtensorflow.so //native_client:deepspeech
```
Comment (lissyx) [inline native_client/README.md]: nit: isn't `//native_client:deepspeech` going to build only the deepspeech client binary, and not the depending libdeepspeech.so ?

Comment (Cwiiis) [inline native_client/README.md]: No, we build the binary with the Makefile - deepspeech corresponds to the library target (so it'll build libdeepspeech.so and libdeepspeech.a)

File: native_client/deepspeech.h (hunk 1 above)
Comment (lissyx) [inline native_client/deepspeech.h]: It is indeed outside of the scope, but reading that part of the code made me think of that. The more we can embed in the model, the better IMHO.

Comment (Cwiiis) [inline native_client/deepspeech.h]: Agreed - and I didn't think when originally replying, but it needn't be in the training model either, we could just embed that part in the exported model. It'd be a shame to have to rewrite all this code (again) to use Tensorflow though...

**Comment by kdavis-mozilla** [comment]
@Cwiiis Could you rebase this again?

There's some strangeness in your PR that includes changes you didn't make into your PR.

For example see master's [DeepSpeech.py#L49](https://github.com/mozilla/DeepSpeech/blob/master/DeepSpeech.py#L49) and your [DeepSpeech.py#L49](https://github.com/Cwiiis/DeepSpeech/blob/issue-557/DeepSpeech.py#L49)

**Comment by Cwiiis** [comment]
Rebased.

File: native_client/BUILD [hunk 10]
```diff
@@ -1,30 +1,30 @@
...
-    srcs = ["c_speech_features/c_speech_features.c"],
-    hdrs = ["c_speech_features/c_speech_features.h",
-            "c_speech_features/c_speech_features_config.h"],
-    includes = ["c_speech_features"],
-    deps = [":kissfft"],
-    copts = ["-std=gnu99"]
+    # Note, we need gnu99 as opposed to c99 as the latter doesn't define M_PI
+    # Also, fma/avx/avx2 enable in gcc cause significant performance decreases in
+    # c_speech_features and so are force-disabled.
+    copts = ["-std=gnu99" ] + if_x86(["-mno-fma", "-mno-avx", "-mno-avx2"]),
```
Comment (lissyx) [inline native_client/BUILD]: nit: multiline and match commands with comments :)

File: native_client/deepspeech.h [hunk 11]
```diff
@@ -4,83 +4,118 @@
...
+                int aSampleRate);
+  };
+
+  /**
+   * @brief Given audio, return a vector suitable for input to a DeepSpeech
+   *        model trained with the given parameters.
+   *
+   * Extracts MFCC features from a given audio signal and adds the appropriate
+   * amount of context to run inference on a DeepSpeech model trained with
+   * the given parameters.
```
Comment (kdavis-mozilla) [inline native_client/deepspeech.h]: Maybe you should comment on the relation between this function and `Model::getInputVector`. This would just help quell any confusion.

File: native_client/deepspeech.cc [hunk 12]
```diff
@@ -13,16 +13,18 @@
 
 using namespace tensorflow;
 
-struct _DeepSpeechPrivate {
+namespace DeepSpeech {
+
+struct _Private {
```
Comment (kdavis-mozilla) [inline native_client/deepspeech.cc]: If you change the forward declaration in the header, this needs to change to.

File: native_client/README.md (hunk 9 above)
Comment (Cwiiis) [inline native_client/README.md]: Documentation error, indeed you need the dash before march/mtune.

File: native_client/Makefile [hunk 13]
```diff
@@ -31,7 +31,7 @@ endif
 
 OS := $(shell uname -s)
 CFLAGS  += -L${TFDIR}/bazel-bin/tensorflow -L${TFDIR}/bazel-bin/native_client $(EXTRA_CFLAGS)
-LDFLAGS += -ltensorflow -ldeepspeech -lc_speech_features -lkissfft $(EXTRA_LDFLAGS)
+LDFLAGS += -ltensorflow -ldeepspeech $(EXTRA_LDFLAGS)
```
Comment (Cwiiis) [inline native_client/Makefile]: Did the command that included building //tensorflow:libtensorflow.so complete successfully? It doesn't move the result to the root of the binary directory unless you build it explicitly, even if it's depended on. If so, I guess I'll have to try multiple paths, but would appreciate the verification first.

(i.e. try `bazel build -c opt --copt=-mtune=native --copt=-march=native --copt=-O3 //tensorflow:libtensorflow.so` and see if it puts libtensorflow.so in the 'right' place)

File: native_client/Makefile [hunk 14]
```diff
@@ -56,15 +56,15 @@ bindings: setup.cfg
 run: deepspeech
 	${META_LD_LIBRARY_PATH}=${TFDIR}/bazel-bin/tensorflow:${TFDIR}/bazel-bin/native_client:${${META_LD_LIBRARY_PATH}} ./deepspeech ${ARGS}
 
+debug: deepspeech
+	${META_LD_LIBRARY_PATH}=${TFDIR}/bazel-bin/tensorflow:${TFDIR}/bazel-bin/native_client:${${META_LD_LIBRARY_PATH}} gdb --args ./deepspeech ${ARGS}
+
 install:
```
Comment (Cwiiis) [inline native_client/Makefile]: The install target only installs the libraries right now - It'd make sense for it to install the binary too though, I'll change that.

File: native_client/Makefile [hunk 15]
```diff
@@ -56,15 +56,15 @@ bindings: setup.cfg
...
 	install -d ${PREFIX}/lib
-	install -m 0644 ${TFDIR}/bazel-bin/tensorflow/libtensorflow.so ${PREFIX}/lib/
-	install -m 0644 ${TFDIR}/bazel-bin/native_client/libkissfft.so ${PREFIX}/lib/
-	install -m 0644 ${TFDIR}/bazel-bin/native_client/libc_speech_features.so ${PREFIX}/lib/
-	install -m 0644 ${TFDIR}/bazel-bin/native_client/libdeepspeech.so ${PREFIX}/lib/
+	install -s -m 0644 ${TFDIR}/bazel-bin/tensorflow/libtensorflow.so ${PREFIX}/lib/
```
Comment (Cwiiis) [inline native_client/Makefile]: (see upper comment)

File: native_client/Makefile [hunk 16]
```diff
@@ -56,15 +56,15 @@ bindings: setup.cfg
...
+	install -s -m 0644 ${TFDIR}/bazel-bin/native_client/libdeepspeech.so ${PREFIX}/lib/
 
 uninstall:
-	rm -f ${PREFIX}/lib/libtensorflow.so
-	rm -f ${PREFIX}/lib/libkissfft.so
-	rm -f ${PREFIX}/lib/libc_speech_features.so
 	rm -f ${PREFIX}/lib/libdeepspeech.so
+	rm -f ${PREFIX}/lib/libtensorflow.so
```
Comment (Cwiiis) [inline native_client/Makefile]: ditto.

File: native_client/deepspeech.h [hunk 17]
```diff
@@ -4,83 +4,118 @@
...
 #include <cstddef>
 
-typedef struct _DeepSpeechPrivate DeepSpeechPrivate;
+namespace DeepSpeech
+{
 
-class DeepSpeech {
-  private:
-    DeepSpeechPrivate* mPriv;
+  typedef struct _Private Private;
```
Comment (Cwiiis) [inline native_client/deepspeech.h]: Because doing the former would require including the Tensorflow headers in the deepspeech header, and those headers aren't publicly available.

File: native_client/deepspeech.h [hunk 18]
```diff
@@ -4,83 +4,118 @@
...
-     *                       (ncep * ncontext) in @p aMFCC.
-     */
-    void getMfccFrames(const short* aBuffer,
-                       unsigned int aBufferSize,
-                       int aSampleRate,
-                       float** aMfcc,
-                       int* aNFrames = NULL,
-                       int* aFrameLen = NULL);
+    public:
+      /**
```
Comment (Cwiiis) [inline native_client/deepspeech.h]: Whoops, good catch.

File: native_client/deepspeech.h [hunk 19]
```diff
@@ -4,83 +4,118 @@
...
+       * @brief Initialise a DeepSpeech context.
+       *
+       * @param aModelPath The path to the frozen model graph.
+       * @param aNCep The number of cepstrum the model was trained with.
+       * @param aNContext The context window the model was trained with.
+       *
+       * @return A DeepSpeech model.
+       */
+      Model(const char* aModelPath, int aNCep, int aNContext);
+      ~Model();
```
Comment (Cwiiis) [inline native_client/deepspeech.h]: I'm not sure what we'd document for it... 'Destroys the model object' seems a bit redundant.

File: native_client/deepspeech.h (hunk 11 above)
Comment (Cwiiis) [inline native_client/deepspeech.h]: Good call.

File: native_client/deepspeech.h (hunk 17 above)
Comment (kdavis-mozilla) [inline native_client/deepspeech.h]: I don't see how it would require anything. It's a forward declaration.

File: native_client/deepspeech.cc (hunk 12 above)
Comment (Cwiiis) [inline native_client/deepspeech.cc]: See above.

File: native_client/deepspeech.h (hunk 19 above)
Comment (kdavis-mozilla) [inline native_client/deepspeech.h]: I was just thinking something along the lines of releases resources held by the Model object. Just so people know it actually has to run to free held resources.

File: native_client/Makefile (hunk 13 above)
Comment (kdavis-mozilla) [inline native_client/Makefile]: The build fails on OS X
```bash
kdaviss-MBP:tensorflow kdavis$ bazel build -c opt --copt=-march=native --copt=-mtune=native --copt=-O3 //tensorflow:libtensorflow.so //native_client:deepspeech
...
ERROR: /Users/kdavis/Code/Cwiiis/tensorflow/native_client/BUILD:9:1: C++ compilation of rule '//native_client:deepspeech' failed: cc_wrapper.sh failed: error executing command external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-march=native' ... (remaining 115 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
error: invalid argument '-std=gnu99' not allowed with 'C++/ObjC++'
INFO: Elapsed time: 1007.991s, Critical Path: 602.48s
```
OS X doesn't seem to support the std gnu99.
